subprojects/core-api/src/main/java/org/gradle/api/file/SourceDirectorySet.java

```diff
@@ -104,27 +108,66 @@
      */
     PatternFilterable getFilter();
 
+    /**
+     * Configure the directory to assemble the compiled classes into.
+     *
+     * @return The destination directory property for this set of sources.
+     * @since 6.1
+     */
+    @Incubating
+    DirectoryProperty getDestinationDirectory();
+
+    /**
+     * Returns the directory property that is bound to the task that produces the output via {@link #compiledBy(TaskProvider, Function)}.
+     * Use this as part of a classpath or input to another task to ensure that the output is created before it is used.
+     *
+     * Note: To define the path of the output folder use {@link #getDestinationDirectory()}
+     *
+     * @return The output directory property for this set of sources.
+     * @since 6.1
+     */
+    @Incubating
+    Provider<Directory> getClassesDirectory();
+
+    /**
+     * Define the task responsible for processing the source.
+     *
+     * @param taskProvider the task responsible for compiling the sources (.e.g. compileJava)
+     * @param mapping a mapping from the task to the task's output directory (e.g. AbstractCompile::getDestinationDirectory)
+     * @since 6.1
+     */
+    @Incubating
+    <T extends Task> void compiledBy(TaskProvider<T> taskProvider, Function<T, DirectoryProperty> mapping);
+
     /**
      * Returns the directory to put the output for these sources.
      *
      * @return The output directory for this set of sources.
+     * @deprecated get the value from {@link #getDestinationDirectory()} instead
```

Comment:
I think most people will want `getClassesDirectory()` if they want to depend on the output

Reply:
Good point. Changing this in follow-up deprecation PR.